kernel: add futex support

wentongwu opened the pull request:
k_futex structure lives in application memory. And when using
    futexes, the majority of the synchronization operations are
    performed in user mode. A user-mode thread employs the futex
    wait system call only when it is likely that the program has
    to block for a longer time until the condition becomes true.
    When the condition comes true, futex wake operation will be
    used to wake up one or more threads waiting on that futex.
    
    This patch implements two futex operations: k_futex_wait and
    k_futex_wake. For k_futex_wait, the comparison with the expected
    value, and starting to sleep are performed atomically to prevent
    lost wake-ups. If different context changed futex's value after
    the calling use-mode thread decided to block himself based on
    the old value, the comparison will help observing the value
    change and will not start to sleep. And for k_futex_wake, it
    will wake at most num_waiters of the waiters that are sleeping
    on that futex. But no guarantees are made on which threads are
    woken, that means scheduling priority is not taken into
    consideration.

    Also add futex test cases.

And the original discussion is here https://github.com/zephyrproject-rtos/zephyr/issues/14493,
And Andrew give me lots of comments at here https://github.com/zephyrproject-rtos/zephyr/pull/15846
And I also borrow some Andrew's code from https://github.com/andrewboie/zephyr/commit/3eaefc4596e62deaa0396814cf715fb08aa235f6, thanks @andrewboie 

   Fixes: #14493.

wentongwu commented:
@andrewboie could you please take a look the patch? Thanks

andrewboie commented on include/kernel.h:
```diff
@@ -2056,6 +2057,99 @@ static inline void *z_impl_k_queue_peek_tail(struct k_queue *queue)
+ * @cond INTERNAL_HIDDEN
+ */
+
+#define _K_FUTEX_INITIALIZER(obj, initial_count) \
```
delete this, it doesn't even work as defined

andrewboie commented on include/kernel.h:
```diff
@@ -2056,6 +2057,99 @@ static inline void *z_impl_k_queue_peek_tail(struct k_queue *queue)
+ *
+ * @return N/A
+ */
+static inline void k_futex_reset(struct k_futex *futex)
```
why do we need this API? remove this

andrewboie commented on include/kernel.h:
```diff
@@ -2056,6 +2057,99 @@ static inline void *z_impl_k_queue_peek_tail(struct k_queue *queue)
+ */
+__syscall int k_futex_wake(struct k_futex *futex, unsigned int num_waiters);
+
+#define K_FUTEX_DEFINE(name, initial_count) \
```
I really don't think we need this

andrewboie commented on tests/kernel/futex/src/main.c:
```diff
@@ -0,0 +1,379 @@
+	k_mem_domain_add_thread(&futex_domain, k_current_get());
+
+	ztest_test_suite(test_futex,
+			 ztest_unit_test(test_futex_wait_forever_wake),
```
shouldn't all of these except test_futex_wait_forever_wake_from_isr() be running in user mode, with ztest_user_unit_test()?

as it stands right now, I don't k_futex_wake() being called from user mode anywhere

wentongwu replied:
yes, the case of k_futex_wake being called from user mode is added

andrewboie commented on tests/kernel/futex/src/main.c:
```diff
@@ -0,0 +1,379 @@
+
+struct k_mem_domain futex_domain;
+K_APPMEM_PARTITION_DEFINE(futex_partition);
+K_APP_BMEM(futex_partition) int timeout;
```
strictly speaking, you could just add these globals to the ztest partition using ZTEST_BMEM.
it might make this test easier to run on systems that only have 8 MPU regions.

andrewboie commented on tests/kernel/futex/src/main.c:
```diff
@@ -0,0 +1,379 @@
+#define TOTAL_THREADS_WAITING (5)
+#define STACK_SIZE (1024 + CONFIG_TEST_EXTRA_STACKSIZE)
+
+#define futex_wake_from_isr(futex) irq_offload(futex_isr_wake, futex)
```
make this a function, not a macro, like you did with futex_isr_wake()

andrewboie commented on tests/kernel/futex/src/main.c:
```diff
@@ -0,0 +1,379 @@
+	case K_FOREVER:
+		zassert_true(ret_value == 0,
+		     "k_futex_wait failed when it shouldn't have");
+		zassert_false(ret_value == 0,
```
why are you checking the same assertion twice?

wentongwu replied:
the first one is check return value of k_futex_wait, the second one is checking no return(this test case is intend to test futex wait only).

andrewboie commented on tests/kernel/futex/src/main.c:
```diff
@@ -0,0 +1,379 @@
+	}
+}
+
+void futex_wait_wake_task(void *p1, void *p2, void *p3)
```
this function is nearly identical to futex_wait_task, it would help to document why they have to be separate, or better yet combine them

andrewboie commented on tests/kernel/futex/src/main.c:
```diff
@@ -0,0 +1,379 @@
+			K_NO_WAIT);
+
+	/* giving time for the futex_wait_task to execute */
+	k_sleep(K_MSEC(200));
```
let's not add sleeps like this. Please note that CONFIG_ZTEST_THREAD_PRIORITY is of the cooperative priority of -1. creating the child thread at the same or higher priority, and then calling k_yield() should do the same thing. you should be able to get rid of most of the k_sleep() calls in this C file.

wentongwu replied:
yes, this is a good idea to make deterministic output. Thanks

andrewboie commented on tests/kernel/futex/prj.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_ZTEST=y
+CONFIG_IRQ_OFFLOAD=y
+CONFIG_TEST_USERSPACE=y
+CONFIG_SMP=n
```
why was this specifically disabled?

wentongwu replied:
the cases so far don't consider SMP.

andrewboie replied:
Remove this line.
We have no SMP enabled userspace targets, and when we do we will need to ensure this works.

wentongwu commented:
Memory region         Used Size  Region Size  %age Used
           FLASH:       76440 B       512 KB     14.58%
            SRAM:         37 KB        32 KB    115.62%
        IDT_LIST:         120 B         2 KB      5.86/opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/../lib     /gcc/arm-zephyr-eabi/8.3.0/../../../../arm-zephyr-eabi/bin/ld: zephyr/zephyr_prebuilt.elf section `priv_stacks_noinit' will not fit in region `SRAM'
/opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/../../../../arm-zephyr-eabi/bin/ld: section .intList VMA [0000000020008000,0000000020008077] overlaps section priv_stacks_noinit VMA [0000000020005400,00000000200093ff]
/opt/sdk/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/../../../../arm-zephyr-eabi/bin/ld: region `SRAM' overflowed by 5120 bytes
collect2: error: ld returned 1 exit status``

wentongwu commented:
The failed two cases are because of not enough RAM size, I re-submit the patch which reduce RAM consuming.

wentongwu commented:
@andrewboie @ioannisg could you please review the updated changes? Thanks.

ioannisg commented on include/kernel.h:
```diff
@@ -2056,6 +2057,76 @@ static inline void *z_impl_k_queue_peek_tail(struct k_queue *queue)
+ * @brief Pend the current thread on a futex
+ *
+ * Tests that the supplied futex contains the expected value, and if so,
+ * goes to sleep until some other thread calls k_futex_wake() on it.
```
Why does it go to sleep? Is it a formal expression to indicate that the thread is moved to pending state?

andrewboie replied:
@ioannisg you should probably review http://man7.org/linux/man-pages/man2/futex.2.html we are using the same semantics as Linux.

ioannisg commented on kernel/CMakeLists.txt:
```diff
@@ -50,6 +50,7 @@ target_sources_ifdef(
   mem_domain.c
   userspace_handler.c
   userspace.c
+  futex.c
```
seems the list is sorted alphabetically...

ioannisg commented on kernel/futex.c:
```diff
@@ -0,0 +1,130 @@
+/*
```
Would it make sense to place all file under CONFIG_USERSPACE?

andrewboie replied:
It is, look at target_sources_ifdef(CONFIG_USERSPACE, ...) in CMakeLists.txt

ioannisg replied:
@andrewboie that's actually what I had checked, so I was confused, because inside this file there are some CONFIG_USERSPACE #ifdefs.

andrewboie replied:
oh I see, yes @wentongwu please remove those

ioannisg commented on kernel/futex.c:
```diff
@@ -0,0 +1,130 @@
+#ifdef CONFIG_USERSPACE
+Z_SYSCALL_HANDLER(k_futex_wake, futex, num_waiters)
+{
+	if (Z_SYSCALL_MEMORY_WRITE(futex, sizeof(struct k_futex)) != 0) {
```
Do we need to check READ access as well, @wentongwu @andrewboie ?

andrewboie replied:
WRITE implies read with that API

ioannisg reviewed (commented):
Mostly, ok. I left a couple of comments/questions.

andrewboie commented on include/kernel.h:
```diff
@@ -2056,6 +2057,76 @@ static inline void *z_impl_k_queue_peek_tail(struct k_queue *queue)
+ * is not taken into consideration.
+ *
+ * @param futex Futex to wake up pending threads
+ * @param num_waiters Max number of threads to wake up
```
There are no use-cases for num_waiters except for 1 and INT_MAX. Windows 10, when they introduced futexes, simplified this to a boolean. Make this a bool "wake_all", if true, wake all pending threads, if false, wake one.

andrewboie commented on tests/kernel/futex/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
```
Please move the futex testcase under tests/kernel/mem_protect

andrewboie commented on include/kernel.h:
```diff
@@ -2056,6 +2057,76 @@ static inline void *z_impl_k_queue_peek_tail(struct k_queue *queue)
+ * @brief Wake a specified number of threads pending on a futex
+ *
+ * Wake up at most num_waiters threads pending on the supplied futex.
+ * No guarantees are made on which threads are woken, scheduling priority
```
Actually, since we are using z_unpend_first_thread() to wake up threads, the highest priority thread *will* be woken up. Change the documentation to reflect this.

andyross commented on kernel/futex.c:
```diff
@@ -0,0 +1,130 @@
+
+	if (timeout == K_NO_WAIT) {
+		k_spin_unlock(&lock, key);
+		return -EBUSY;
```
Even if it's not pathological, this is the wrong return value.  EBUSY is intended to be for for hardware or subsystems that is doing something else.  This is the call returning synchronously exactly as requested.  I'd return a zero here, if you absolutely must have this case in the code.

andrewboie replied:
How about returning -ETIMEDOUT, since technically it did.

andyross commented on kernel/futex.c:
```diff
@@ -0,0 +1,130 @@
+
+	ret = z_pend_curr(&lock, key, thread_q, timeout);
+	if (ret == -EAGAIN) {
+		ret = -ETIMEDOUT;
```
I see what you're doing here: you want separate return values for "we raced on the expected value, try again" and "the timeout expired".

Nonetheless this is confusing.  In fact everywhere else in zephyr where a timeout expires, the relevant call returns -EAGAIN.    In fact this is the default behavior of the pend API in the scheduler, for exactly this reason.  We should be consistent here.  I know the error says "timed out", but that's not the error we use for timeouts, basically. :)

If you need to disambiguate the cases, maybe returning -EINVAL for the "expected has changed" case would be a better choice.  The argument is, in fact, "invalid" for the case where we want to wait.

andrewboie replied:
I asked Wentong to return -ETIMEDOUT because to match the return values from Linux.

```
       ETIMEDOUT
              The  operation in futex_op employed the timeout specified in timeout, and the timeout
              expired before the operation completed.
```

> We should be consistent here. I know the error says "timed out", but that's not the error we use for timeouts, basically. :)

Why is this the case? Why don't these APIs return -ETIMEDOUT? We can break APIs for 2.0, maybe we should do that?

I'm trying to keep the return values consistent with Linux to make it simple to port futex-using userspace code to Zephyr.

andyross replied:
I wouldn't cry if we changed the timeout return.  I agree it's a cleaner naming.  It would be a ton of code to change in all the tests though.  Not it. :)

andyross reviewed (changes requested):
Some nitpicking.  "The internal tracking object should be a futex-specific struct and not a bare wait queue" is the only one I feel strongly about.

andrewboie commented on kernel/futex.c:
```diff
@@ -0,0 +1,130 @@
+}
+#endif
+
+int z_impl_k_futex_wake(struct k_futex *futex, unsigned int num_waiters)
```
agreed, I asked Wentong to change this to a boolean for wake-one/wake-all

andrewboie commented on scripts/gen_kobject_list.py:
```diff
@@ -170,6 +172,15 @@ def write_gperf_table(fp, eh, objs, static_begin, static_end):
 
+    num_futexes = eh.get_futex_counter()
+    if (num_futexes != 0):
+        fp.write("static _wait_q_t futex_wait_q[%d] = {\n" % num_futexes)
```
good idea

wentongwu replied:
yes, that's good, I'm changing the code.

andrewboie commented on kernel/futex.c:
```diff
@@ -0,0 +1,130 @@
+		return -EAGAIN;
+	}
+
+	if (timeout == K_NO_WAIT) {
```
if we made it an assert we would still need to do a runtime check from the syscall, I'd rather we just left it.

andyross replied:
Actually thinking on this some more: just remove the whole case.  The K_NO_WAIT will be passed through to the pend call below, which will return -EAGAIN synchronously and we'll return.  No need to check this at all.

wentongwu replied:
yeah, thanks

andrewboie commented on kernel/futex.c:
```diff
@@ -0,0 +1,130 @@
+#include <init.h>
+#include <ksched.h>
+
+static struct k_spinlock lock;
```
That seems a reasonable change.

wentongwu replied:
There should be enough memory if enabled user space. But I don't think there is lock breaking for single core system.  And I will add separate spinlock for each futex object because we support SMP.

andyross commented on kernel/futex.c:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	if (woken) {
+		z_reschedule(&lock, key);
```
To elaborate here: there **is** a behavior difference between the two strategies: in one, a cooperative caller will only yield if it actually woke up threads on the futex.  In the other, it will **also** yield if anything else in the system happened to become runnable.

I just argue the latter case is too subtle and dangerous to be worth documenting for app authors and that we should just say "k_futex_wait is a reschedule point". And if we do that, we should pick the simpler implementation here.

wentongwu replied:
agree, thanks

wentongwu commented on include/kernel.h:
```diff
@@ -2056,6 +2057,76 @@ static inline void *z_impl_k_queue_peek_tail(struct k_queue *queue)
+ * k_futex are tracked as kernel objects and can live in user memory
+ * so any access bypasses the kernel object permission management
+ * mechanism.
+ */
```
thanks

ioannisg reviewed (approved):
Looks ok to me

wentongwu committed:
kernel: add futex support

A k_futex is a lightweight mutual exclusion primitive designed
to minimize kernel involvement. Uncontended operation relies
only on atomic access to shared memory. k_futex structure lives
in application memory. And when using futexes, the majority of
the synchronization operations are performed in user mode. A
user-mode thread employs the futex wait system call only when
it is likely that the program has to block for a longer time
until the condition becomes true. When the condition comes true,
futex wake operation will be used to wake up one or more threads
waiting on that futex.

This patch implements two futex operations: k_futex_wait and
k_futex_wake. For k_futex_wait, the comparison with the expected
value, and starting to sleep are performed atomically to prevent
lost wake-ups. If different context changed futex's value after
the calling use-mode thread decided to block himself based on
the old value, the comparison will help observing the value
change and will not start to sleep. And for k_futex_wake, it
will wake at most num_waiters of the waiters that are sleeping
on that futex. But no guarantees are made on which threads are
woken, that means scheduling priority is not taken into
consideration.

Fixes: #14493.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
tests: add futex test cases

add futex test cases.

Fixes: #14493.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu commented:
@andrewboie @andyross could you please review the new patch? Thanks.

wentongwu commented:
@andrewboie @andyross  could you please review the new patch? Thanks

andyross reviewed (approved):
Looks good.  Still not loving the inconsistency with EAGAIN/ETIMEDOUT, but hobgoblins and all...
